Update checklist when tasks are completed

meisekimiu opened the pull request:
Right now the checklist functions well, but users need to refresh the page when they complete tasks to seem them checked off. This PR updates the checklist so it automatically refreshes when an item may be checked off. It does this by refreshing whenever an event is sent to the AnalyticsService.

**Steps to test:**
1. Create a new account with the `hideChecklist` field set to true.
2. Complete an action
3. The checklist should update whenever a relevant task is completed.

**Considerations before release**
This behavior was mainly tested on a local environment without any latency. We might want to add a slight delay before refreshing the checklist to account for network latency.

liam-lloyd commented on src/app/core/services/edit/edit.service.ts:
```diff
@@ -463,6 +465,15 @@ export class EditService {
+          entityId: this.accountService.getAccount()?.accountId.toString(),
+          version: 1,
+          body: {
+            noTransmit: true,
```
This pattern, where we send an "analytics" event that doesn't actually go to any analytics repository in order to let the checklist know that it should refresh, doesn't seem quite right to me. This might just be a naming issue; an events service that sends a subset of its events to analytics seems reasonable to me. But does it makes sense for the client to have an internal event saying we tried to move a record and a separate event fired on the back end saying we actually did move the record? Maybe, I'm not sure. Perhaps we should discuss this in sprint planning on Monday.

meisekimiu commented:
**Big refactoring done!**

The Analytics / Mixpanel Services have undergone a huge refactoring. Here's a quick summary of the changes:
- The services are now renamed to the EventService and AnalyticsService respectively.
- The event data type passed to the EventService is more abstract and compact; now a caller of `EventService.dispatch` does not know whether the event is being sent to the `event` endpoint or not. It is just raising an abstract event. This abstract event is caught by the AnalyticsService which adapts the abstract event to the more verbose event object that the `event` endpoint is expecting. No other outside code knows about this more verbose type.
- The new `PermanentEvent` type limits which actions and entities can be paired together. So now a combination like `{entity: 'record', action: 'open_archive_profile'}` is an invalid type. Additionally, each type of event takes in a whole Account, Record, LegacyContact, etc. object instead of just taking in an `entityId` property.
- The `DataService` is now provided in the `AppModule` since it is a dependency of the AnalyticsService and the AnalyticsService is initialized in the AppModule.
- The checklist observer now waits one second before refreshing the checklist.

Open things to discuss:
- Currently the `PermanentEvent` type is made using a type union, where each underlying type identifies itself through the `entity` property. To get the `entityId` to send to the `event` endpoint, the AnalyticsService needs to use a `switch` statement on the `entity` property so it can narrow the `PermanentEvent` type and recover specific attached objects. Technically it is probably better to use polymorphic dispatch on this (some kind of abstract method like `PermanentEvent.getEntityId()` can be called and each subclass defines its own implementation). However, I also like the idea  of events being very simple "plain old data structures", and using polymorphism means that constructors have to be called every time we create a new event. I am fine changing this to be polymorphic, or just keeping it as it is for now.
- The delay functionality I added might not be the most ideal way of doing this. We could technically wait for the `event` endpoint to fully finish calling, but I think that would involve having another event on the `EventService` that is dispatched once all of the main `EventObserver.update` calls are finished. What do we think about that?
- I have to fix some merge conflicts which were pretty much inevitable with a refactoring this big.

meisekimiu commented:
Additionally: While all of the "uncovered" calls to the EventService that codecov reports are still technically uncovered, the actual calls to the EventService that have been added are covered in the AnalyticsService's integration tests.

liam-lloyd reviewed (approved):
Just a few comments on some variable names, otherwise looks good

meisekimiu commented on src/app/app.module.ts:
```diff
@@ -43,17 +43,18 @@ import {
-  analyticsService: AnalyticsService,
-  mixpanelService: MixpanelService,
+  analyticsService: EventService,
+  mixpanelService: AnalyticsService,
```
Yep, thanks for catching.

meisekimiu commented on src/app/onboarding/components/create-new-archive/create-new-archive.component.ts:
```diff
@@ -47,7 +47,7 @@ export class CreateNewArchiveComponent implements OnInit, OnDestroy {
   @Input() pendingArchive: ArchiveVO;
 
-  private mixpanelActions: { [key: string]: MixpanelAction } = {
+  private mixpanelActions: { [key: string]: AccountEventAction } = {
```
Renamed this to `actionsForScreen`.

meisekimiu commented on src/app/shared/services/analytics/analytics.service.ts:
```diff
@@ -1,35 +1,141 @@
-
-export interface AnalyticsObserver {
-  update(eventData: EventData): Promise<void>;
+export interface MixpanelData {
```
Renamed this to `EventRequestBody`.
